components/server/src/scm/scm-service.ts

```diff
@@ -0,0 +1,234 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { inject, injectable } from "inversify";
+import { Authorizer } from "../authorization/authorizer";
+import { Config } from "../config";
+import { TokenProvider } from "../user/token-provider";
+import { CommitContext, Project, SuggestedRepository, Token, User, WorkspaceInfo } from "@gitpod/gitpod-protocol";
+import { HostContextProvider } from "../auth/host-context-provider";
+import { RepoURL } from "../repohost";
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
+import { AuthProviderService } from "../auth/auth-provider-service";
+import { UserService } from "../user/user-service";
+import { GitTokenScopeGuesser } from "../workspace/git-token-scope-guesser";
+import { ApplicationError, ErrorCodes } from "@gitpod/gitpod-protocol/lib/messaging/error";
+import {
+    SuggestedRepositoryWithSorting,
+    sortSuggestedRepositories,
+    suggestionFromProject,
+    suggestionFromRecentWorkspace,
+    suggestionFromUserRepo,
+} from "../workspace/suggested-repos-sorter";
+
+@injectable()
+export class ScmService {
+    constructor(
+        @inject(Config) protected readonly config: Config,
+        @inject(Authorizer) private readonly auth: Authorizer,
+        @inject(TokenProvider) private readonly tokenProvider: TokenProvider,
+        @inject(HostContextProvider) private readonly hostContextProvider: HostContextProvider,
+        @inject(AuthProviderService) private readonly authProviderService: AuthProviderService,
+        @inject(UserService) private readonly userService: UserService,
+        @inject(GitTokenScopeGuesser) private readonly gitTokenScopeGuesser: GitTokenScopeGuesser,
+    ) {}
+
+    public async getToken(userId: string, query: { host: string }): Promise<Token> {
```

Comment:
Do you think it makes sense to add scm-service.spec.ts to test new functions? I think not if they are moved mostly as is without changing the logic. Just asking to double check.

Reply:
They were moved. But still some basic tests would be good.  👍🏻